### PR TITLE
Validator and parser

### PR DESCRIPTION
**Pull request by misslecter**

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+export const getPersonalNumberParts = (value: string): PersonalNumberParts => {
+  let firstPart = '';
+  let secondPart = '';
+
+  const parts = value.split('/');
+  if (parts.length === 1) {
+    firstPart = value.substr(0, 6);
+    secondPart = value.substr(6);
+  } else {
```
Comment (tomaskourim) [inline src/utils.ts:19]: no a kdyz bude parts.length=== 12 tak jsme s tim ok?

File: src/utils.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  }
+
+  return { firstPart, secondPart };
+};
+
+export const getAge = (dateOfBirth: Date): number => {
+  const ageDifMs = Date.now() - dateOfBirth.getTime();
+  const ageDate = new Date(ageDifMs);
+  const year = 1970;
+  return Math.abs(ageDate.getUTCFullYear() - year);
```
Comment (tomaskourim) [inline src/utils.ts:31]: fakt jo? I pro lidi narozene pred 1970? I pro lidi narozene BC?

File: src/validator.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
+import { getPersonalNumberParts, personalNumberAddingTwentyIssueYear, tenDigitPersonalNumberIssueYear, unprobableMonthAddition, womanMonthAddition } from './utils';
+
+const validateDate = (year: number, month: number, day: number): boolean => {
+  const fullYear = year >= tenDigitPersonalNumberIssueYear ? 1900 + year : 2000 + year;
```
Comment (tomaskourim) [inline src/validator.ts:4]: tohle je bud zbytecne nebo nefunguje podle me

File: src/parser.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  if (month > womanMonthAddition) {
+    gender = Gender.F;
+    month -= womanMonthAddition;
+  }
+
+  if (month > unprobableMonthAddition) {
+    month -= unprobableMonthAddition;
+  }
+
+  const fullYear = year >= tenDigitPersonalNumberIssueYear || secondPart.length === 3 ? 1900 + year : 2000 + year;
```
Comment (tomaskourim) [inline src/parser.ts:46]: tohle v roce 2054 prestane fungovat. Spis se bere, ze jsou lide mladsi 100 let, ale zalezi na kontextu taky

File: src/__tests__/validator.test.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import { validate } from '../validator';
+
+test('Valid personal numbers with a slash', () => {
```
Comment (tomaskourim) [inline src/__tests__/validator.test.ts:3]: urcite bych pridal min 1 test na ruzne edge case - prictena 20, zenske, 9mistne RC

File: src/utils.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
+export interface PersonalNumberParts {
+  firstPart: string;
+  secondPart: string;
+}
+
+export const personalNumberAddingTwentyIssueYear = 4;
```
Comment (tomaskourim) [inline src/utils.ts:6]: asi vim co myslis, ale v kodu to nikde neni (nebo nevidim)

**Review by tomaskourim — CHANGES_REQUESTED** [review]
Par veci. Repo bych pojmenoval personal-number-parser. Udajne existuji nejaci divni polaci z Ostravy a okoli, co maji 11mistne polske RC. Asi na ne muzem kaslat.

File: src/__tests__/validator.test.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+});
+
+test('Valid personal numbers without a slash', () => {
+  expect(validate('7209097280')).toBe(true);
+});
+
+test('Invalid personal numbers with invalid length', () => {
+  expect(validate('')).toBe(false);
+  expect(validate('72090/7280')).toBe(false);
+  expect(validate('720909/728')).toBe(false);
```
Comment (marazt) [inline src/__tests__/validator.test.ts:14]: jak pise @tomaskourim  - tohle cislo muze byt ok - starsi lidi meli maji jen 3 cisla za lomitkem.

File: src/parser.ts (hunk 4 above)
Comment (marazt) [reply src/parser.ts:46]: Jo, na tohle by chtel dat nejaky opsn, ktery rekne, jak se to prave ma chovat pro tyhle extra stary lidi.

**Review by marazt — COMMENTED** [review]
Dotaz - tohle jsme prece davali do Kontlin kodu ne? Nestaci to ztoho opsat, pokud se tak nestalo?

**Comment by misslecter** [comment]
@marazt opsala jsem tvůj kod z covid vaxxu, je stejný jako ten v kotlinu?

File: src/utils.ts (hunk 1 above)
Comment (misslecter) [reply src/utils.ts:19]: pro pripad `1234/5678/1234/5678/1234/5678/1234/5678` bychom skoncili jen s prvni dvojici `1234` a `5678`, ktera by pak neprosla dalsi validaci. Ale pro jistotu upravim :)

File: src/utils.ts (hunk 2 above)
Comment (misslecter) [reply src/utils.ts:31]: je to nejaka epocha pouzivana v JS. Nasla jsem ale jinou verzi vypoctu, ktera by mela byt vic accurate, tak pouziju ji.

File: src/__tests__/validator.test.ts (hunk 7 above)
Comment (misslecter) [reply src/__tests__/validator.test.ts:14]: tohle by znamenalo, ze je ten clovek narozen v roce 1872 a takove lide jsme nepodporovali. Ale uz podporujem :)

**Comment by misslecter** [comment]
@tomaskourim @marazt Otázka je jak moc staré lidi chceme podporovat. Například, pro ženu narozenou 18.05.2000, rodné číslo `005518/532` by nemělo být validní, protože nesplnuje format. Vrácíme ale 121 letou ženu. Je to takhle v pohodě?

File: src/utils.ts (hunk 6 above)
Comment (misslecter) [reply src/utils.ts:6]: https://github.com/mild-blue/personal-number/blob/2dd29536fddd185ae8b6de57752f728389dc12e3/src/validator.ts#L53

File: src/parser.ts (hunk 4 above)
Comment (misslecter) [reply src/parser.ts]: opraveno tak, aby fungovalo i po roce 2054 i pred rokem 1900

File: src/validator.ts (hunk 3 above)
Comment (misslecter) [reply src/validator.ts]: opraveno, tak snad spravne

File: src/validator.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
+import { getFullBirthYear, getPersonalNumberParts, personalNumberAddingTwentyIssueYear, unprobableMonthAddition, womanMonthAddition } from './utils';
+
+const validateDate = (year: number, month: number, day: number): boolean => {
+  try {
+    const _ = new Date(year, month - 1, day);
```
Comment (marazt) [inline src/validator.ts:5]: nestaci jen `new Date(year, month - 1, day);` ?

**Comment by LukasForst** [comment]
kdyz uz jsme u jmena (cc @tomaskourim ) pojmenoval bych to `czech-personal-number` protoze tohle je skutecne jen ceska vec

`parser` bych nedaval, protoze tohle umi i validovat 😉

File: src/__tests__/parser.test.ts [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    age: getAge(dateOfBirth),
+    gender,
+    birthOrder,
+    dateOfBirth,
+    controlDigit
+  };
+
+  expect(parse(`${firstPart}0${birthOrder}${controlDigit}`)).toEqual(expectedResult);
+  expect(parse(`${firstPart}/0${birthOrder}${controlDigit}`)).toEqual(expectedResult);
+  expect(parse(`${firstPart}/${birthOrder}`)).toEqual(undefined);
```
[the remark this answers is not in the document]
Comment (misslecter) [reply src/__tests__/parser.test.ts]: jo jo, a melo by to byt ok. Ma to tak i jiny validator:

image

[the remark this answers is not in the document]
Comment (misslecter) [reply src/__tests__/parser.test.ts]: pridala jsem na to explicitni test

File: src/__tests__/parser.test.ts [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+    gender,
+    birthOrder,
+    dateOfBirth,
+    controlDigit
+  };
+
+  expect(parse(`${firstPart}/${birthOrder}`)).toEqual(expectedResult);
+});
+
+test('A female born on 01.01.1953 with addition of 20', () => {
```
[the remark this answers is not in the document]
Comment (misslecter) [reply src/__tests__/parser.test.ts]: opravila jsem to

File: src/validator.ts [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+      return false;
+    }
+  }
+
+  if (month > womanMonthAddition) {
+    month -= womanMonthAddition;
+  }
+
+  if (month > unprobableMonthAddition) {
+    if (year >= personalNumberAddingTwentyIssueYear) {
```
[the remark this answers is not in the document]
Comment (misslecter) [reply src/validator.ts]: opraveno

File: src/parser.ts [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+ * @returns PersonalNumberParseResult if value is a valid personal number, undefined if value is not a valid personal number.
+ */
+export const parse = (value: string): PersonalNumberParseResult | undefined => {
+  if (!validate(value)) {
+    return undefined;
+  }
+
+  const { firstPart, secondPart } = getPersonalNumberParts(value);
+
+  const year = Number(firstPart.substr(0, 2));
```
[the remark this answers is not in the document]
Comment (misslecter) [reply src/parser.ts]: refaktorovano

File: src/__tests__/parser.test.ts [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+
+  const expectedResult: PersonalNumberParseResult = {
+    age: getAge(dateOfBirth),
+    gender,
+    birthOrder,
+    dateOfBirth,
+    controlDigit
+  };
+
+  expect(parse(`${firstPart}0${birthOrder}${controlDigit}`)).toEqual(expectedResult);
```
[the remark this answers is not in the document]
Comment (misslecter) [reply src/__tests__/parser.test.ts]: `birthOrder` je `number` a nemuze zacitat 0, proto je soucasti stringu

File: src/parser.ts [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+  let month = Number(firstPart.substr(2, 2));
+  const day = Number(firstPart.substr(4, 2));
+  const birthOrder = Number(secondPart.substr(0, 3));
+  const controlDigit = secondPart.length === 4 ? Number(secondPart.substr(3, 1)) : undefined;
```
[the remark this answers is not in the document]
Comment (misslecter) [reply src/parser.ts:36]: refaktorovano tak, aby veskera logika byla v jedne funkci

File: src/utils.ts [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+export const unprobableMonthAddition = 20;
+
+export const getPersonalNumberParts = (value: string): PersonalNumberParts => {
+  let firstPart = '';
+  let secondPart = '';
+
+  const parts = value.split('/');
+  if (parts.length === 1) {
+    firstPart = value.substr(0, 6);
+    secondPart = value.substr(6);
```
[the remark this answers is not in the document]
Comment (misslecter) [reply src/utils.ts:11]: vratim prazdny string, delku overim v `parse` funkci

**Review by JanecekPetr — COMMENTED** [review]
(I'm a random lurker trying to refer some friends to Mild Blue, but since this is a public repo, I took a look :) )

In particular, how is this better than / different from https://github.com/kub1x/rodnecislo?

File: src/model/Gender.enum.ts [hunk 16]
```diff
@@ -0,0 +1,4 @@
+export enum Gender {
```
[the remark this answers is not in the document]
Comment (kubantjan) [reply src/model/Gender.enum.ts:1]: > (I'm a random lurker trying to refer some friends to Mild Blue, but since this is a public repo, I took a look :) )
> 
> In particular, how is this better than / different from https://github.com/kub1x/rodnecislo?

@misslecter do we know this repo?

File: src/parser.ts [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+      result: undefined,
+      message: 'Personal number value cannot be empty.'
+    };
+  }
+
+  const { firstPart, secondPart } = splitPersonalNumberValue(value);
+  if (firstPart.length !== 6) {
+    return {
+      result: undefined,
+      message: `First part of personal number must have 6 digits. Given first part: ${firstPart}.`
```
Comment (tomaskourim) [inline src/parser.ts:26]: Tady a v tom elsu bych vratil i cely ten input. Protože když ve splitPersonalNumberValue bude vycházet délka >2, tak to vrátí ve firstPart i secondPart prázdný string - to uživatele dost zmate

File: src/parser.ts [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+    const concatenated = Number(firstPart + secondPart);
+
+    const moduloElevenOk = concatenated % 11 === 0;
+    const withoutLastDigit = concatenated / 10;
+    const moduloTenOk = withoutLastDigit % 11 === 10 && controlDigit === 0;
+
+    if (!moduloTenOk && !moduloElevenOk) {
+      return {
+        result: undefined,
+        message: `Second part does not satisfy modulo condition. Given second part: ${secondPart}.`
```
Comment (tomaskourim) [inline src/parser.ts:49]: taky bych vracel celé RČ a text třeba
Given personal number does not satisfy modulo condition.

File: src/parser.ts [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+  let gender: Gender = Gender.M;
+  if (month > womanMonthAddition) {
+    gender = Gender.F;
+    month -= womanMonthAddition;
+  }
+
+  // Get year
+  const year = getFullBirthYear(secondPart, Number(firstPart.substr(0, 2)));
+
+  // Remove unprobableMonthAddition if there is one
```
Comment (tomaskourim) [inline src/parser.ts:67]: improbable

File: README.md [hunk 20]
```diff
@@ -1 +1,2 @@
-# personal-number
\ No newline at end of file
+# Personal number
+Validates and extracts information from Czech personal identity number
```
Comment (tomaskourim) [inline README.md:2]: Nechceme sem něco napsat. Jakože jak to počítá, pro jaké roky to funguje atd.?

File: src/parser.ts (hunk 19 above)
Comment (tomaskourim) [reply src/parser.ts:67]: na spoustě míst

File: src/parser.ts [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+
+  // Get day
+  const day = Number(firstPart.substr(4, 2));
+
+  // Get date of birth
+  const dateOfBirth = getDateOfBirth(year, month, day);
+  if (!dateOfBirth) {
+    return {
+      result: undefined,
+      message: `No valid date of birth can be created with values: year = ${year}, month = ${month} and day = ${day}.`
```
Comment (tomaskourim) [inline src/parser.ts:87]: `Not a valid date of birth. Values: ...`
Protoze kdyz to je v budoucnosti, tak muzes vytvorit datum narozeni, ale neni validni

File: src/parser.ts [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+    };
+  }
+
+  // Get birth order
+  const birthOrder = Number(secondPart.substr(0, 3));
+
+  // Get age
+  const age = getAge(dateOfBirth);
+  if (age < 0) {
```
Comment (tomaskourim) [inline src/parser.ts:96]: to uz kontrolujes v getDateOfBirth. Nicmene tam bych to nekontroloval a nechal to az zde (a ignoroval predchozi komentar)

File: src/__tests__/parser.test.ts [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+
+  const expectedResult: ParsedPersonalNumber = {
+    age: getAge(dateOfBirth),
+    gender,
+    birthOrder,
+    dateOfBirth,
+    controlDigit
+  };
+
+  const secondPart = `0${birthOrder}${controlDigit}`;
```
Comment (tomaskourim) [inline src/__tests__/parser.test.ts:21]: muzeme se zbavit techto nahodne vkladanych nul? vsude prosim

**Review by tomaskourim — CHANGES_REQUESTED** [review]
uz jen drobnosti
